File: dsa-problems/Strings/solutions/javascript/STR-010-balanced-brackets-limited-skips.js
function canBalanceWithSkips(s, k) {
  let balance = 0;
  let skipsUsed = 0;

  for (let char of s) {
    if (char === "(") {
      balance++;
    } else {
      // char === ')'
      balance--;
      if (balance < 0) {
        // Need to skip this ')'
        skipsUsed++;
        balance = 0;
      }
    }
  }

  // Remaining balance are unmatched '('
  const totalSkipsNeeded = skipsUsed + balance;
  return totalSkipsNeeded <= k;
}


const readline = require('readline');
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
let tokens = [];
rl.on('line', (line) => { tokens.push(...line.trim().split(/\s+/)); });
rl.on('close', () => {
    if(tokens.length===0) return;
    let ptr = 0;
    const s = tokens[ptr++];
    const k = parseInt(tokens[ptr++]);
    const sol = new Solution();
    console.log(sol.canBalanceWithSkips(s, k));
});

File: dsa-problems/Strings/solutions/java/STR-010-balanced-brackets-limited-skips.java
import java.util.*;

class Solution {
    public boolean canBalanceWithSkips(String s, int k) {
        int balance = 0;
        int skipsUsed = 0;

        for (char c : s.toCharArray()) {
            if (c == '(') {
                balance++;
            } else {  // c == ')'
                balance--;
                if (balance < 0) {
                    // Need to skip this ')'
                    skipsUsed++;
                    balance = 0;
                }
            }
        }

        // Remaining balance are unmatched '('
        int totalSkipsNeeded = skipsUsed + balance;
        return totalSkipsNeeded <= k;
    }
}

















class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String s = sc.next();
        int k = sc.nextInt();
        Solution sol = new Solution();
        System.out.println(sol.canBalanceWithSkips(s, k));
        sc.close();
    }
}
